### ISSUE TITLE
Support accepting socket from system manager instead of managing own socket

### ISSUE THREAD

**Issue by lucaswerkmeister** [keybase/client#3508]
Currently, the Keybase service insists on managing its socket itself; when passed an existing file, it attempts to delete the “stale” socket, and exits if that fails.

In the interest of managing Keybase with systemd (comments in `run_keybase` as well as in this issue tracker indicate that this is intended to be supported at some point), the service should also have an option to accept an open socket from the system manager (systemd). The best documentation for adding socket support to a program seems to be [these](http://0pointer.de/blog/projects/socket-activation.html) [two](http://0pointer.net/blog/projects/socket-activation2.html) blog posts (`systemd.socket(5)` and `systemd.service(5)` focus more on the unit file, not on what the service has to do). In C, this seems to be the gist of it:
```c
#include "sd-daemon.h"

int fd;
assert (sd_listen_fds(0) == 1);
fd = SD_LISTEN_FDS_START + 0;
assert (sd_is_socket(fd, AF_UNSPEC, SOCK_STREAM, 1));
```
There also appears to be a [Go systemd binding](https://github.com/coreos/go-systemd), including an [example](https://github.com/coreos/go-systemd/tree/master/examples/activation/httpserver) for socket-based activation.

Note: macOS’ launchd apparently also supports socket-based activation. I’m not sure how similar the mechanism is.

**Comment by maxtaco** [comment]
@oconnor663 can say more, but we'd love to let `systemd` deal with a lot of stuff we've been dealing with (like log rotation, socket management, daemon startup and restart as an unprivileged user), but we previously didn't think we'd cover enough of the Linuxes.  @oconnor663 we were waiting for 16.04 LTS right, so maybe we can think about switching over?

**Comment by lucaswerkmeister** [comment]
Well, for now you could also add the option to the service and not ship unit files yet – leave it to adventurous users to tinker with systemd support :)

(Also, you might want to follow [this bug](https://bugs.freedesktop.org/show_bug.cgi?id=73809) – currently, a `keybase.mount` user unit seems to be impossible.)

**Comment by oconnor663** [comment]
Yes, we should be able to add a flag like `--socket-fd=3`, and then systemd unit files that want to use socket activation can just hardcode that. I'll file a task.

There are two issues I'm worried about before I'd want to make our official packages depend on systemd:

1. Pre-systemd Ubuntu versions are still supported by Canonical. I don't have numbers on how many people are still using 14.04 or (G-d help us) 12.04. I wouldn't want to break those people unless we could be sure they were a low single digit percentage of our users. (Adding some more detailed version logging could give us the pie chart for _current_ users, but I bet old Ubuntu versions are underrepresented among people who wait months to try beta crypto software :p)
2. I don't know how well systemd user services interact with the X11/Wayland desktop. Do they get `DISPLAY` and all the other relevant environment variables? The [Arch wiki's section on this](https://wiki.archlinux.org/index.php/Systemd/User#Xorg_and_systemd) makes it sound like there are a lot of complications.

The XDG autostart directory solves both of those problems for us right now. That said, people trying to use Keybase on a headless server might benefit from a semi-official systemd way of doing things. (Note that the KBFS experience on headless machines is currently very rough. There's nothing that can pop up at you to ask you for credentials or to confirm someone's broken tracking statement, so _lots_ of operations will just hang or error out without giving you clear instructions.)

I'm totally unfamiliar with how systemd handles filesystem mounts. But presumably it's possible to write a regular user unit that happens to run FUSE commands on the inside, even if systemd doesn't know anything about it? I guess you'd lose the magical start-on-first-use feature, but if you autostarted it on login maybe it would mostly work?

**Comment by lucaswerkmeister** [comment]
Re 2: On my system, user units seem to have a pretty full environment, including `DISPLAY`. I suspect that depends on the display manager, though (I use gdm).

But I’m not sure it matters, actually. I would start the keybase service and mount the keybase filesystem via systemd – those can happen without a display – and leave the actual desktop application in XDG autostart, which is where it belongs IMO.

I believe it’s possible to write a user *service* unit that runs `kbfsfuse` and thus happens to mount the keybase filesystem (though systemd won’t know anything about the mount point). It’s not possible to write a proper *mount* unit because systemd currently always mounts those with `mount -t`, and `mount` doesn’t allow the `-t` option if you’re not root. [FreeDesktop Bug 73809](https://bugs.freedesktop.org/show_bug.cgi?id=73809), which I linked above, might fix that.

**Comment by Wolf480pl** [comment]
You could also make the deamon take the socket as stdin, which would make it compatible with systemd (intetd-style) socket activation, but without making it depend on libsystemd.

Additionaly, it would also make it compatible with other socket activation implementations, eg. [s6-ipcserver-socketbinder](http://www.skarnet.org/software/s6/s6-ipcserver-socketbinder.html)

Anyway, I hope you don't add a hard dependency on libsystemd. An optional dependency (or evencopy-pasting the socket activation bits of libsystemd, like rsyslog does) would be better in my opinion.

**Comment by oconnor663** [comment]
We're in the middle of shipping systemd support, see [here](https://github.com/keybase/client/issues/5355#issuecomment-342590835).
